### PR TITLE
kafka: update the kafka to 2.3.0

### PR DESCRIPTION
**Pull request by zmalik**
This PR

- bumps Apache Kafka to 2.3.0.
- make livenessProbe and readinessProbe less aggressive.
- Adds CPU/Memory limits for the stateful pods
- Updates the default value of Zookeeper URI
- Add more sections of documentations around configuration/updates and upgrades

And also introduces the `latest` docs folder that when ready for release can be bumped to the versioned docs folder like `0.1.1.` or `0.2.0` etc

### CODE REVIEW

File: repository/kafka/docs/latest/configuration.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Configuration 
+
+### Resources
+
+By default, KUDO Kafka resource configuration is set to be minimal optimal for the production usage.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
By default, KUDO Kafka resource configuration is set to the minimum recommended values for production usage. 
```

File: repository/kafka/docs/latest/configuration.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# Configuration 
+
+### Resources
+
+By default, KUDO Kafka resource configuration is set to be minimal optimal for the production usage. 
+But users can tune the requirements depending on high or low workload of the Kafka cluster.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
But users can and should tune the configurations based on the workload requirements of their Kafka cluster.  
```

File: repository/kafka/docs/latest/configuration.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+### Resources
+
+By default, KUDO Kafka resource configuration is set to be minimal optimal for the production usage. 
+But users can tune the requirements depending on high or low workload of the Kafka cluster.  
+
+##### Tuning the resources for the Kafka Cluster
+
+```
+kubectl kudo install kafka --instance=my-kafka-name \
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md:11]: Are users able to run this multiple times with different settings and Kudo will automatically know to adjust the current deployment? Or is this only valid for initial deployment?

File: repository/kafka/docs/latest/configuration.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            -p NUM_IO_THREADS=10
+            -p NUM_NETWORK_THREADS=5
+          
+```
+
+##### Storage
+
+By default, the Kafka operator will use the default storage class of the Kubernetes cluster. 
+
+To deploy the Kafka cluster using a storage class that isn't the default, you can use the parameter `STORAGE_CLASS`
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
To deploy Kafka using a different storage class, you can use the parameter `STORAGE_CLASS`
```

File: repository/kafka/docs/latest/configuration.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+kubectl kudo install kafka --instance=my-kafka-name -p STORAGE_CLASS=<STORAGE_CLASS_NAME>
+```
+
+##### Deploying an ephemeral Kafka cluster without persistence
+
+```
+kubectl kudo install kafka --instance=my-kafka-name -p PERSISTENT_STORAGE=false
+```
+
+Having `PERSISTENT_STORAGE` value `false` means that any data or logs in broker will be lost after the a pod restart or rescheduling.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
Having `PERSISTENT_STORAGE` value `false` means that any data or logs inside the brokers will be lost after a pod restart or rescheduling.
```

Comment (samvantran) [reply repository/kafka/docs/latest/configuration.md]: Should this setting even be allowed? Sounds like a place for future trouble

File: repository/kafka/docs/latest/configuration.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+Deploying without persistent storage isn't recommended for production usage. 
+
+##### Metrics
+
+By default, the Kafka cluster will have the JMX Exporter enabled. You can check more information around how KUDO Kafka exposes metrics in [monitoring](./monitoring.md) documentation.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
By default, the Kafka cluster will have the JMX Exporter enabled. You can check more information around how KUDO Kafka exposes metrics in [monitoring](./monitoring.md).
```

File: repository/kafka/docs/latest/configuration.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+##### Zookeeper PATH
+
+By default, the Kafka cluster will try to use a zookeeper path same as the instance name.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]: This is a little confusing. What path will it try to use by default? `/zookeeper` ?

File: repository/kafka/docs/latest/configuration.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+User can override the default zk path by using `ZOOKEEPER_PATH` parameter.
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
To override the default zk path, use the `ZOOKEEPER_PATH` parameter.
```

File: repository/kafka/docs/latest/configuration.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+By default, KUDO Kafka resource configuration is set to be minimal optimal for the production usage. 
+But users can tune the requirements depending on high or low workload of the Kafka cluster.  
+
+##### Tuning the resources for the Kafka Cluster
+
+```
+kubectl kudo install kafka --instance=my-kafka-name \
+            -p ZOOKEEPER_URI=zk-zk-0.zk-hs:2181,zk-zk-1.zk-hs:2181,zk-zk-2.zk-hs:2181 \
+            -p ZOOKEEPER_PATH=/custom-path \
+            -p BROKER_CPUS=3000m
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]: None of these additional params need `\`?

File: repository/kafka/docs/latest/configuration.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+By default, the Kafka cluster will have the JMX Exporter enabled. You can check more information around how KUDO Kafka exposes metrics in [monitoring](./monitoring.md) documentation.
+
+##### Zookeeper PATH
+
+By default, the Kafka cluster will try to use a zookeeper path same as the instance name.
+
+User can override the default zk path by using `ZOOKEEPER_PATH` parameter.
+
+```
+kubectl kudo install kafka --instance=my-kafka-name -p ZOOKEEPER_URI=zk-zk-0.zk-hs:2181,zk-zk-1.zk-hs:2181,zk-zk-2.zk-hs:2181 -p ZOOKEEPER_PATH=/custom-path
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
kubectl kudo install kafka --instance=my-kafka-name \
  -p ZOOKEEPER_URI=zk-zk-0.zk-hs:2181,zk-zk-1.zk-hs:2181,zk-zk-2.zk-hs:2181 \
  -p ZOOKEEPER_PATH=/custom-path
```
Easier to identify the `ZOOKEEPER_PATH`

File: repository/kafka/docs/latest/configuration.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+```
+
+##### Docker image
+
+The [Dockerfile](https://github.com/mesosphere/dcos-kafka-service/blob/master/images/Dockerfile) used in the KUDO Kafka operator is hosted in the [dcos-kafka-service](https://github.com/mesosphere/dcos-kafka-service) repo
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md]:
```suggestion
The Dockerfile used to build the KUDO Kafka operator is hosted in the [dcos-kafka-service](https://github.com/mesosphere/dcos-kafka-service/blob/master/images/Dockerfile) repo
```

File: repository/kafka/docs/latest/install.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
+# Installing KUDO Kafka operator
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
# Installing the KUDO Kafka Operator
```

File: repository/kafka/docs/latest/install.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+Requirements:
+
+- [Install KUDO controller](https://kudo.dev/docs/getting-started/)
+- [Install KUDO cli](https://kudo.dev/docs/cli/)
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
- Install the [KUDO CLI](https://kudo.dev/docs/cli/)
```

File: repository/kafka/docs/latest/install.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
+# Installing KUDO Kafka operator
+
+Requirements:
+
+- [Install KUDO controller](https://kudo.dev/docs/getting-started/)
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
- Install the [KUDO controller](https://kudo.dev/docs/getting-started/)
```

File: repository/kafka/docs/latest/install.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+- [Install KUDO cli](https://kudo.dev/docs/cli/)
+
+
+## Installing the Operator
+
+#### Install the Zookeeper
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
#### Install Zookeeper 
```

File: repository/kafka/docs/latest/install.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+
+## Installing the Operator
+
+#### Install the Zookeeper 
+```
+kubectl kudo install zookeeper --instance=zk
+```
+
+#### Install the Kafka
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
#### Install Kafka 
```

File: repository/kafka/docs/latest/install.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+kubectl kudo plan status --instance=kafka
+Plan(s) for "kafka" in namespace "default":
+.
+└── kafka (Operator-Version: "kafka-0.1.1" Active-Plan: "kafka-deploy-177524647")
+    └── Plan deploy (serial strategy) [COMPLETE]
+        └── Phase deploy-kafka (serial strategy) [COMPLETE]
+            └── Step deploy (COMPLETE)
+```
+
+You can check all [configuration options](./configuration.md)
```
Comment (samvantran) [inline repository/kafka/docs/latest/install.md]:
```suggestion
You can view all configuration options [here](./configuration.md)
```

File: repository/kafka/docs/latest/limitations.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
+
+
+# Limitations
+
+
+
+Below is the list of parameters that can only be configured during bootstrap time.
```
Comment (samvantran) [inline repository/kafka/docs/latest/limitations.md]:
```suggestion
Below is a list of parameters that can only be configured during bootstrap time:
```

File: repository/kafka/docs/latest/limitations.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+**DISK_SIZE**, **STORAGE_CLASS** and **PERSISTENT_STORAGE** 
+
+These next storage related parameters cannot be changed. Resizing disk, changing storage class or switching between pesistent storage or ephemeral using parameters isn't supported.
```
Comment (samvantran) [inline repository/kafka/docs/latest/limitations.md]:
```suggestion
These storage-related parameters cannot be changed after initial deployment. Repeat: using parameters to resize disk, change storage class, or switch between persistent/ephemeral storage is not supported.
```

File: repository/kafka/docs/latest/limitations.md [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+
+
+### Resizing the PVC 
+
+Resizing the disk depends on the storage class of the Kubernetes.
```
Comment (samvantran) [inline repository/kafka/docs/latest/limitations.md]:
```suggestion
Resizing the disk depends on the storage class of your Kubernetes cluster.
```

File: repository/kafka/docs/latest/limitations.md [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+
+
+To resize the disk being used by a broker. Users can edit the `pvc` and expand the disk
```
Comment (samvantran) [inline repository/kafka/docs/latest/limitations.md]:
```suggestion
To resize the disk being used by a broker, users can edit the `pvc` and expand the disk.
```

File: repository/kafka/docs/latest/limitations.md [hunk 22]
```diff
@@ -0,0 +1,21 @@
+
+
+# Limitations
+
+
+
```
Comment (samvantran) [inline repository/kafka/docs/latest/limitations.md]: Can we reduce all the 3-line-breaks to 1? The rest of your docs follow this pattern (and it's easier to read in code-form)

File: repository/kafka/docs/latest/monitoring.md [hunk 23]
```diff
@@ -0,0 +1,38 @@
+# Monitoring
+
+KUDO Kafka operator comes by default the JMX Exporter agent enabled.
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
By default, the KUDO Kafka Operator comes with the JMX Exporter agent enabled. 
```

File: repository/kafka/docs/latest/monitoring.md [hunk 24]
```diff
@@ -0,0 +1,38 @@
+# Monitoring
+
+KUDO Kafka operator comes by default the JMX Exporter agent enabled. 
+
+When Kafka operator deployed with parameter `METRICS_ENABLED=true` (which defaults to `true`) then:
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
When the Kafka operator is deployed with the parameter `METRICS_ENABLED=true` (which defaults to `true`) then:
```

File: repository/kafka/docs/latest/monitoring.md [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+
+- Each broker bootstraps with [JMX Exporter](https://github.com/prometheus/jmx_exporter) java agent exposing the metrics at `9094/metrics`
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
- Each broker bootstraps with the [JMX Exporter](https://github.com/prometheus/jmx_exporter) java agent exposing the metrics at `9094/metrics`
```

File: repository/kafka/docs/latest/monitoring.md [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+```
+kubectl describe svc kafka-svc
+...
+Port:              metrics  9094/TCP
+TargetPort:        9094/TCP
+...
+```
+
+### Using Prometheus Service Monitor
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
### Using the Prometheus Service Monitor
```

File: repository/kafka/docs/latest/monitoring.md [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+To use the prometheus service monitor, its necessary to have installed the prometheus operator previously in the cluster.
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
To use the prometheus service monitor, it's necessary to have installed the prometheus operator previously in the cluster.
```

File: repository/kafka/docs/latest/monitoring.md [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+
+If the Kafka Cluster Service in the default namespace, we can use the next example of the service-monitor.
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
If Kafka is using the default namespace, we can create a service-monitor with the following:
```

File: repository/kafka/docs/latest/monitoring.md [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+```
+kubectl create -f resources/service-monitor.yaml
+```
+
+Install the [grafana dashboard](./resources/grafana-dashboard.json), and we should be able to see the Kafka dashaboard.
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
Upload this [grafana json](./resources/grafana-dashboard.json) and you should see the following Kafka dashboard:
```

File: repository/kafka/docs/latest/monitoring.md [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+
+![Grafana Dashboards](./resources/grafana-capture.png)
+
+### Disable JMX Exporter
```
Comment (samvantran) [inline repository/kafka/docs/latest/monitoring.md]:
```suggestion
### Disable the JMX Exporter
```

File: repository/kafka/docs/latest/release-notes.md [hunk 31]
```diff
@@ -0,0 +1,5 @@
+
+- Apache Kafka upgraded to 2.3.0.
```
Comment (samvantran) [inline repository/kafka/docs/latest/release-notes.md:2]: Should you put a date?

File: repository/kafka/docs/latest/update.md [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+
+Check the limitations to see which parameters can only be set during bootstrap time. 
+
+
+
+## Scaling the brokers
+
+For the stateful workload it is not recommended to configure the HPA or VPA for Kafka brokers. 
+
+More controlled scaling is more conveninent considering the nature of stateful workload.
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
It is recommended that users closely monitor and control broker scaling due to the nature of stateful workloads.
```

File: repository/kafka/docs/latest/update.md [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+
+#### Tuning the configuration 
+
+Check the limitations to see which parameters can only be set during bootstrap time. 
+
+
+
+## Scaling the brokers
+
+For the stateful workload it is not recommended to configure the HPA or VPA for Kafka brokers.
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
Users should not configure the HPA or VPA for Kafka brokers. 
```

File: repository/kafka/docs/latest/update.md [hunk 34]
```diff
@@ -0,0 +1,137 @@
...
+    Namespace:    default
+    UID:          d34d38e7-1940-4670-83a3-56b486bccff4
+  Status:         COMPLETE
+Events:
+  Type    Reason               Age                 From                      Message
+  ----    ------               ----                ----                      -------
+  Normal  CreatePlanExecution  5m3s (x2 over 20h)  instance-controller       Creating "deploy" plan execution
+  Normal  PlanCreated          5m3s                instance-controller       PlanExecution "kafka-deploy-351387000" created
+  Normal  PlanComplete         3m56s               planexecution-controller  PlanExecution kafka-deploy-351387000 completed
+```
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]: [This whole section](https://github.com/kudobuilder/operators/pull/50/files#diff-5d5fa4bb1f28111b62dd2ea37b56dba2R21-R57) of code does not seem relevant to the example of scaling brokers from 3-5. I think you could delete it.

File: repository/kafka/docs/latest/update.md [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+kafka-kafka-2   1/1     Running   0          8m53s
+kafka-kafka-3   1/1     Running   0          10m
+kafka-kafka-4   1/1     Running   0          9m21s
+```
+
+
+
+**Vertically** 
+
+To scale vertically the brokers statefulset pods we can use the parameters also.
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
To scale vertically, we can update the broker's statefulset.
```

File: repository/kafka/docs/latest/update.md [hunk 36]
```diff
@@ -0,0 +1,137 @@
...
+
+## Scaling the brokers
+
+For the stateful workload it is not recommended to configure the HPA or VPA for Kafka brokers. 
+
+More controlled scaling is more conveninent considering the nature of stateful workload.  
+
+#### Horizontally 
+
+To scale horizontally we need to scale our brokers count from default 3 to 5.
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
To scale horizontally, we can increase the broker count.
```

File: repository/kafka/docs/latest/update.md [hunk 37]
```diff
@@ -0,0 +1,137 @@
...
+[ ... lines removed for clarity ...]
+    Requests:
+      cpu:      500m
+      memory:   2048m
+[ ... lines removed for clarity ...]
+```
+
+
+
+Lets change the cpu request from `500m` to `700m` and double the memory request from `2048` to `4096` and not forget the limits, that cannot be lower than the requested resources.
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
Let's increase the cpu request from `500m` to `700m` and double the memory request from `2048` to `4096`. Also important is increasing the limits as they cannot be lower than the requested resources. 
```

File: repository/kafka/docs/latest/update.md [hunk 38]
```diff
@@ -0,0 +1,137 @@
...
+      memory:   2048m
+[ ... lines removed for clarity ...]
+```
+
+
+
+Lets change the cpu request from `500m` to `700m` and double the memory request from `2048` to `4096` and not forget the limits, that cannot be lower than the requested resources. 
+
+```
+kubectl patch instance kafka -p '{"spec":{"parameters":{"BROKER_CPUS":"700m", "BROKER_MEM":"4096m", "BROKER_CPUS_LIMIT":"3000m", "BROKER_MEM_LIMIT":"6144"}}}' --type=merge
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]: You're passing `BROKER_MEM: 4096m` but `BROKER_MEM_LIMIT: 6144`. Is this valid? The `m` for limit is implied? I would still add it to maintain consistency with the example.

File: repository/kafka/docs/latest/update.md [hunk 39]
```diff
@@ -0,0 +1,137 @@
...
+```
+
+Now we will see a rolling upgrade of the pods towards
```
Comment (samvantran) [inline repository/kafka/docs/latest/update.md]:
```suggestion
This will initiate a rolling upgrade of the pods to a new statefulset.
```

File: repository/kafka/docs/latest/upgrade.md [hunk 40]
```diff
@@ -0,0 +1,77 @@
+# Upgrading the KUDO Kafka Operator
+
+
+
+KUDO upgrades work by linking the Kafka cluster Instance object to the correct operatorVersion object.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
KUDO Kafka upgrades work by linking the Kafka cluster `Instance` object to the correct `operatorVersion` object.
```

File: repository/kafka/docs/latest/upgrade.md [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+
+We can have multiple operator versions in the same Kubernetes cluster. 
+One instance object represents a Kafka Cluster. And different instances can use same or different operator versions.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]: Did you mean to add a line break or keep it as part of the line above? You'll have to add another line-break if you want it on new line.

Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
One instance object represents a Kafka Cluster and multiple instances can use the same or different operator versions.
```

File: repository/kafka/docs/latest/upgrade.md [hunk 42]
```diff
@@ -0,0 +1,77 @@
...
+
+![operator-upgrade-1](./resources/images/operator-upgrade-1.png)
+
+The upgrade process can be done by patching the Kafka Instance that holds all the configuration around the Kafka cluster, and pointing it to the a new OperatorVersion object.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
Upgrading can be done by patching the Kafka Instance that holds all the configuration of a specific Kafka cluster and pointing it to a new OperatorVersion object.
```

File: repository/kafka/docs/latest/upgrade.md [hunk 43]
```diff
@@ -0,0 +1,77 @@
...
+
+
+
+![operator-upgrade-1](./resources/images/operator-upgrade-2.png)
+
+Lets do the upgrade process only having one Kafka cluster and one Kafka operator version.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
The following upgrade procedure assumes one running Kafka cluster and one Kafka operator version already installed (version `kafka-0.1.1`):
```

File: repository/kafka/docs/latest/upgrade.md [hunk 44]
```diff
@@ -0,0 +1,77 @@
...
+
+The upgrade process can be done by patching the Kafka Instance that holds all the configuration around the Kafka cluster, and pointing it to the a new OperatorVersion object.
+
+
+
+![operator-upgrade-1](./resources/images/operator-upgrade-2.png)
+
+Lets do the upgrade process only having one Kafka cluster and one Kafka operator version.  
+
+First we need to install the new operator version
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
Install a new operator version:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 45]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+kubectl kudo install kafka --version=0.2.0 --skip-instance
+
+operator.kudo.k8s.io/kafka unchanged
+operatorversion.kudo.k8s.io/v1alpha1/kafka-0.2.0 created
+```
+Verify if the operator version 0.2.0 is installed. Now we have two operator versions installed.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
Now there are two operator versions installed:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 46]
```diff
@@ -0,0 +1,77 @@
...
+Verify if the operator version 0.2.0 is installed. Now we have two operator versions installed. 
+```
+kubectl  get operatorversions.kudo.k8s.io
+
+NAME              AGE
+kafka-0.1.1       2d2h
+kafka-0.2.0       2m6s
+```
+
+And check the instances that are installed.
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
Check for running instances:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 47]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+kubectl get instances
+NAME           AGE
+kafka-fc6vzn   29h
+```
+
+And we can check the plan status of our Kafka cluster `kafka-fc6vzn`
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
We can check the plan status of our Kafka cluster `kafka-fc6vzn` with:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 48]
```diff
@@ -0,0 +1,77 @@
...
+```
+kubectl kudo plan status --instance=kafka-fc6vzn
+Plan(s) for "kafka-fc6vzn" in namespace "default":
+.
+└── kafka-fc6vzn (Operator-Version: "kafka-0.1.1" Active-Plan: "kafka-fc6vzn-deploy-414458000")
+    └── Plan deploy (serial strategy) [COMPLETE]
+        └── Phase deploy-kafka (serial strategy) [COMPLETE]
+            └── Step deploy (COMPLETE)
+```
+Note the operator version `kafka-0.1.1`
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
**Note:** the operator version is `kafka-0.1.1`
```

File: repository/kafka/docs/latest/upgrade.md [hunk 49]
```diff
@@ -0,0 +1,77 @@
...
+
+Lets update the Kafka cluster from version 0.1.1 to 0.2.0
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
To update the Kafka cluster from version `0.1.1` to `0.2.0`:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 50]
```diff
@@ -0,0 +1,77 @@
...
+
+```
+> kubectl patch instance kafka  -p '{"spec":{"operatorVersion":{"name":"kafka-0.2.0"}}}' --type=merge
+instance.kudo.k8s.io/kafka patched
+```
+
+We can check the plan status
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
Check the plan status again:
```

File: repository/kafka/docs/latest/upgrade.md [hunk 51]
```diff
@@ -0,0 +1,77 @@
...
+kubectl kudo plan status --instance=kafka
+Plan(s) for "kafka" in namespace "default":
+.
+└── kafka (Operator-Version: "kafka-0.2.0" Active-Plan: "kafka-deploy-575561000")
+    └── Plan deploy (serial strategy) [COMPLETE]
+        └── Phase deploy-kafka (serial strategy) [COMPLETE]
+            └── Step deploy (COMPLETE)
+```
+
+Note the operator-version : kafka-0.2.0, means that we have successfully upgraded the operator
```
Comment (samvantran) [inline repository/kafka/docs/latest/upgrade.md]:
```suggestion
**Note:** the operator-version is now `kafka-0.2.0` meaning the Kafka Operator has successfully been upgraded.
```

File: repository/kafka/docs/latest/versions.md [hunk 52]
```diff
@@ -0,0 +1,5 @@
+### Versions
+
+- Apache Kafka 2.12-2.3.0
+- OpenJDK 8
+- JMX Exporter 0.1.0
```
Comment (samvantran) [inline repository/kafka/docs/latest/versions.md:5]: Worth mentioning the KUDO version here too or no? (we mention the SDK version we use in all of our packages)

**Review by samvantran — COMMENTED** [review]
Only had editing comments but overall looks great! Thanks for putting all these docs together. This is huge for users!

File: repository/kafka/docs/latest/configuration.md (hunk 5 above)
Comment (zmalik) [reply repository/kafka/docs/latest/configuration.md]: If there is no `storageclass` in a Kubernetes cluster. Users cannot play with `KUDO Kafka`

File: repository/kafka/docs/latest/configuration.md (hunk 7 above)
Comment (zmalik) [reply repository/kafka/docs/latest/configuration.md]: The Kafka instance name. By default `KUDO` generates an instance name. But users can specify using `--instance` flag.

File: repository/kafka/docs/latest/configuration.md (hunk 3 above)
Comment (zmalik) [reply repository/kafka/docs/latest/configuration.md:11]: custom settings are saved in the `instance` object. So once installed you cannot re-install but you can update the settings. [update docs](https://github.com/kudobuilder/operators/blob/d0a944961fedaf199a13814cd1ea92e1168273cc/repository/kafka/docs/latest/update.md) have information on how to do

File: repository/kafka/docs/latest/configuration.md (hunk 9 above)
Comment (zmalik) [reply repository/kafka/docs/latest/configuration.md]: let me fix that!

File: repository/kafka/docs/latest/release-notes.md (hunk 31 above)
Comment (zmalik) [reply repository/kafka/docs/latest/release-notes.md:2]: When releasing yes, right now its in `latest` directory. So when doing the release the directory should moved to its versioned name directory and add the date also

File: repository/kafka/docs/latest/versions.md (hunk 52 above)
Comment (zmalik) [reply repository/kafka/docs/latest/versions.md:5]: It doesn't use the KUDO as we use SDK in DC/OS.
Different versions of KUDO can deploy the same version of KUDO Kafka

File: repository/kafka/docs/latest/limitations.md (hunk 22 above)
Comment (zmalik) [reply repository/kafka/docs/latest/limitations.md]: done!

File: repository/kafka/README.md [hunk 53]
```diff
@@ -7,3 +7,12 @@ The KUDO Kafka operator creates, configures and manages [Apache Kafka](https://k
...
 
+For the latest master branch you can check  [docs](./docs/latest) docs 
+
+
+## Version Chart
+
+| KUDO Kafka Version | Apache Kafka Version |
+| ------------------ | -------------------- |
+| 0.1.1              | 2.2.1                |
+| master             | 2.3.0                |
```
Comment (alenkacz) [inline repository/kafka/README.md]: should we rather say latest here? or rather explicitly say 0.2.0 ?

File: repository/kafka/docs/latest/configuration.md (hunk 7 above)
Comment (alenkacz) [reply repository/kafka/docs/latest/configuration.md]: maybe we could state that explicitly, that it refers to whatever you provided in `--instance` or what was generated for you, which you can figure out by `kubectl get instances`

File: repository/kafka/docs/latest/monitoring.md [hunk 54]
```diff
@@ -0,0 +1,38 @@
+# Monitoring
+
+By default, the KUDO Kafka Operator comes with the JMX Exporter agent enabled. 
+
+When the Kafka operator is deployed with the parameter `METRICS_ENABLED=true` (which defaults to `true`) then:
+
+- Each broker bootstraps with the [JMX Exporter](https://github.com/prometheus/jmx_exporter) java agent exposing the metrics at `9094/metrics`
+- Adds a port named `metrics` to the Kafka Service
+- Adds a label `kubeaddons.mesosphere.io/servicemonitor: "true"` for the service monitor discovery.
```
Comment (alenkacz) [inline repository/kafka/docs/latest/monitoring.md]: kind of weird seeing kubeaddons here since it's non-public thing...

**Review by alenkacz — APPROVED** [review]
LGTM! Very nice docs

File: repository/kafka/docs/latest/monitoring.md (hunk 54 above)
Comment (zmalik) [reply repository/kafka/docs/latest/monitoring.md]: good catch! it is actually `kudo.dev/servicemonitor: "true"`

File: repository/kafka/README.md (hunk 53 above)
Comment (zmalik) [reply repository/kafka/README.md]: `latest` over `0.2.0` as we can install the operator as it is in master. Which may not be the last stable release.

File: repository/kafka/docs/latest/configuration.md (hunk 7 above)
Comment (zmalik) [reply repository/kafka/docs/latest/configuration.md]: I have added now both cases. Which pretty much explain what would be the `Zookeeper Path` in case we set the parameter and when we don't set the parameter

File: repository/kafka/docs/latest/configuration.md (hunk 3 above)
Comment (samvantran) [reply repository/kafka/docs/latest/configuration.md:11]: Right, what I was getting at was that a user would have to run `patch`, not `install`. So they can't really "tune" the cluster this way. I'll suggest a change in the header.

File: repository/kafka/docs/latest/configuration.md [hunk 55]
```diff
@@ -0,0 +1,70 @@
+# Configuration 
+
+### Resources
+
+By default, KUDO Kafka resource configuration is set to the minimum recommended values for production usage. 
+But users can and should tune the configurations based on the workload requirements of their Kafka cluster.  
+
+##### Tuning the resources for the Kafka Cluster
```
Comment (samvantran) [inline repository/kafka/docs/latest/configuration.md:8]:
```suggestion
##### Deploying Kafka with Parameters
```

File: repository/kafka/docs/latest/versions.md (hunk 52 above)
Comment (samvantran) [reply repository/kafka/docs/latest/versions.md:5]: but I thought upgrades didn't work until kudo 0.3.3 came out?

File: repository/kafka/operator/operator.yaml [hunk 56]
```diff
@@ -22,4 +32,15 @@ plans:
...
             tasks:
-              - deploy
\ No newline at end of file
+              - deploy
+  # this plan is triggered when some parameter present in limitations is updated
+  # https://github.com/kudobuilder/operators/blob/master/repository/kafka/docs/latest/limitations.md
+  not-allowed:
+    strategy: serial
+    phases:
+    - name: not-allowed
```
Comment (samvantran) [inline repository/kafka/operator/operator.yaml:41]: shouldn't KUDO have something built in for this? It would be better if we could mark params as `cannot-be-updated` or something and KUDO would take care of either running this plan or returning an error?

Comment (samvantran) [reply repository/kafka/operator/operator.yaml:41]: rather than having each operator having to implement their own `not-allowed` kind of plan

Comment (zmalik) [reply repository/kafka/operator/operator.yaml:41]: https://kubernetes.slack.com/archives/CG3HTFCMV/p1564063181112400

yeah it will be removed once KUDO provides us some solution. Like a fix for https://github.com/kudobuilder/kudo/issues/652

Comment (zmalik) [reply repository/kafka/operator/operator.yaml:41]: But as for now I don't want users to run into a statefulset with errors.

**Review by samvantran — APPROVED** [review]
LGTM 👍

**Comment by zmalik** [comment]
> but I thought upgrades didn't work until kudo 0.3.3 came out?

that warning should be on release notes of KUDO right? cc: @alenkacz

**Review by fabianbaier — APPROVED** [review]
